packages/types/src/Address.ts

```diff
@@ -31,37 +31,36 @@ export default class Address extends Base<AccountId | AccountIndex> {
     } else if (Array.isArray(value)) {
       return Address.decodeAddress(u8aToU8a(value));
     } else if (isU8a(value)) {
-      if (value[0] === 0xff) {
+      // This allows us to instantiate an address with a raw publicKey. Do this first before
+      // we checking the first byte, otherwise we may split an already-existent valid address
+      if (value.length === 32) {
+        return new AccountId(value);
+      } else if (value[0] === 0xff) {
         return new AccountId(value.subarray(1));
       }
-      return value.length === 32
-        ? new AccountId(value)
-        : new AccountIndex(value);
+
+      const [offset, length] = AccountIndex.readLength(value);
+
+      return new AccountIndex(value.subarray(offset, offset + length), length);
```

Comment:
I still think that the 2nd argument (maxLength) is not needed.

```
// // event: balances.NewAccount(AccountId, AccountIndex, NewAccountOutcome)
// 0100
// 37e027d776cd005c12bcf6722421374a9037167a0ceaf918f341c4ad68d54e59
// 0700000000 // AccountIndex+NewAccountOutcome
```

You pass in `new AccountIndex([07, 00, 00, 00, 00])` (without 2nd arg). That ones decodes it, it finds out that the length is 4 bytes, its encodedLength is 4. Then move on, NewAccountOutcome, 1 byte, and everything works.

Reply:
Yes, however - AccountIndex it can be up to 8 bytes in length. If not added, it would trim those, e.g. 

https://github.com/polkadot-js/api/blob/master/packages/types/src/AccountIndex.spec.ts#L49

https://github.com/polkadot-js/api/blob/master/packages/types/src/Address.spec.ts#L71